runtime/compiler/optimizer/J9TransformUtil.cpp

```diff
@@ -2625,3 +2625,188 @@ J9::TransformUtil::refineMethodHandleLinkTo(TR::Compilation* comp, TR::TreeTop*
+   const char *srcObjTypeSig = src->getSymbolReference() ? src->getSymbolReference()->getTypeSignature(srcSigLen) : 0;
+   const char *destObjTypeSig = dest->getSymbolReference() ? dest->getSymbolReference()->getTypeSignature(destSigLen) : 0;
+
+   // Case 3
+   bool srcArrayCheckNeeded, destArrayCheckNeeded;
+   if (!srcObjTypeSig)
+      srcArrayCheckNeeded = true;
+   else
+      {
+      TR_OpaqueClassBlock *srcClass = comp->fej9()->getClassFromSignature(srcObjTypeSig, srcSigLen, src->getSymbolReference()->getOwningMethod(comp));
+      srcArrayCheckNeeded = srcClass == NULL ||
+                              srcClass == comp->getObjectClassPointer() ||
+                              TR::Compiler->cls.isInterfaceClass(comp, srcClass);
+      }
+
+   if (!destObjTypeSig)
+      destArrayCheckNeeded = true;
+   else
+      {
+      TR_OpaqueClassBlock *destClass = comp->fej9()->getClassFromSignature(destObjTypeSig, destSigLen, dest->getSymbolReference()->getOwningMethod(comp));
+      destArrayCheckNeeded = destClass == NULL ||
+                              destClass == comp->getObjectClassPointer() ||
+                              TR::Compiler->cls.isInterfaceClass(comp, destClass);
+      }
+
+   // Case 2 & 3
+   bool srcAdjustmentNeeded = srcArrayCheckNeeded || srcObjTypeSig[0] == '[';
+   bool destAdjustmentNeeded = destArrayCheckNeeded || destObjTypeSig[0] == '[';
+
+   TR::TransformUtil::separateNullCheck(comp, arrayCopyTT);
+
+   if (!(srcAdjustmentNeeded || destAdjustmentNeeded))
+      {
+      TR::TransformUtil::convertUnsafeCopyMemoryCallToArrayCopyWithSymRefLoad(comp, arrayCopyTT, NULL, NULL);
+      return;
+      }
+
+   // Anchor nodes
+   for (int32_t i=1; i < arraycopyNode->getNumChildren(); i++)
+      {
+      TR::Node* childNode = arraycopyNode->getChild(i);
+      if ( !(childNode->getOpCode().isLoadConst() || childNode->getSymbolReference()->getSymbol()->isAutoOrParm()) )
+         arrayCopyTT->insertBefore(TR::TreeTop::create(comp, TR::Node::create(TR::treetop, 1, arraycopyNode->getChild(i))));
+      }
+
+   TR::CFG *cfg = comp->getFlowGraph();
+   TR::Block *currentBlock = arrayCopyTT->getEnclosingBlock();
+   TR::Block *callBlock = currentBlock->split(arrayCopyTT, cfg, true);
+   TR::Block *nextBlock = callBlock->split(arrayCopyTT->getNextTreeTop(), cfg, true);
+
+   TR::SymbolReference *adjustSrcTempRef = NULL;
+   TR::SymbolReference *adjustDestTempRef = NULL;
+   if (srcAdjustmentNeeded)
+      {
+      adjustSrcTempRef = comp->getSymRefTab()->createTemporary(comp->getMethodSymbol(), TR::Address);
+      adjustSrcTempRef->getSymbol()->setNotCollected();
+      TR::Node* storeNode = TR::Node::createStore(adjustSrcTempRef, src);
+      TR::TreeTop* storeTree = TR::TreeTop::create(comp, storeNode);
+      currentBlock->getExit()->insertBefore(storeTree);
+      callBlock = TR::TransformUtil::insertUnsafeCopyMemoryArgumentChecksAndAdjustForOffHeap(comp, src, adjustSrcTempRef, callBlock, srcArrayCheckNeeded, cfg);
```

Comment:
Instead of `src` this should pass `arraycopyNode->getChild(1)`, which if necessary should have been replaced with a load of a temp by now